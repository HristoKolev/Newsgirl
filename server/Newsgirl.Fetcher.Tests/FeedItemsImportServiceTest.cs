using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using NSubstitute;
using Xunit;

using Newsgirl.Fetcher.Tests.Infrastructure;
using Newsgirl.Shared.Data;
using Newsgirl.Shared.Infrastructure;

namespace Newsgirl.Fetcher.Tests
{
    public class FeedItemsImportServiceTest : DatabaseTest
    {
        [Fact]
        public async Task GetFeedsForUpdate_Returns_All_Feeds()
        {
            var log = Substitute.For<ILog>();

            var importService = new FeedItemsImportService(this.Db, DbConnection, log);

            var feeds = Enumerable.Range(1, 10)
                .Select(i => new FeedPoco
                {
                    FeedHash = i,
                    FeedName = $"feed {i}",
                    FeedUrl = $"url {i}",
                }).ToList();

            await this.Db.BulkInsert(feeds);
            
            var resultFeeds = await importService.GetFeedsForUpdate();
            
            Snapshot.Match(resultFeeds);
        }
        
        
        [Fact]
        public async Task ImportItems_Copies_Items_Correctly()
        {
            var log = Substitute.For<ILog>();

            var importService = new FeedItemsImportService(this.Db, this.DbConnection, log);

            var feeds = Enumerable.Range(1, 10)
                .Select(i => new FeedPoco
                {
                    FeedName = $"feed {i}",
                    FeedUrl = $"url {i}",
                }).ToList();

            foreach (var feed in feeds)
            {
                await Db.Insert(feed);
            }

            var updates = new List<FeedUpdateModel>
            {
                new FeedUpdateModel
                {
                    NewItems = Enumerable.Range(1, 10)
                        .Select(i => new FeedItemPoco
                        {
                            FeedID = 1,
                            FeedItemDescription = $"desc {i}",
                            FeedItemHash = i,
                            FeedItemTitle = $"title {i}",
                            FeedItemUrl = $"url {i}",
                            FeedItemAddedTime = TestHelper.Date2000,
                        }).ToList(),
                    Feed = feeds.First(x => x.FeedID == 1),
                    NewFeedHash = 1,
                },
                new FeedUpdateModel
                {
                    NewItems = Enumerable.Range(100, 10)
                        .Select(i => new FeedItemPoco
                        {
                            FeedID = 2,
                            FeedItemDescription = $"desc {100 + i}",
                            FeedItemHash = i,
                            FeedItemTitle = $"title {100 + i}",
                            FeedItemUrl = $"url {100 + i}",
                            FeedItemAddedTime = TestHelper.Date2000,
                        }).ToList(),
                    Feed = feeds.First(x => x.FeedID == 2),
                    NewFeedHash = 2,
                },
                new FeedUpdateModel
                {
                    Feed = feeds.First(x => x.FeedID == 3), 
                },
                new FeedUpdateModel
                {
                    Feed = feeds.First(x => x.FeedID == 4),
                    NewItems = new List<FeedItemPoco>(),
                    NewFeedHash = 4
                }
            };
            
            await importService.ImportItems(updates.ToArray());

            var resultFeeds = Db.Poco.Feeds.OrderByDescending(x => x.FeedID).ToList();
            var resultFeedItems = Db.Poco.FeedItems.OrderByDescending(x => x.FeedItemID).ToList();
            
            Snapshot.Match(new {resultFeeds, resultFeedItems});
        }
    }
}