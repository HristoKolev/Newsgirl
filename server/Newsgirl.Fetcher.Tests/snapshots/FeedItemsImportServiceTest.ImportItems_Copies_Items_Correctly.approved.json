{
  "resultFeedItems": [
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 209",
      "FeedItemHash": 109,
      "FeedItemID": 20,
      "FeedItemTitle": "title 209",
      "FeedItemUrl": "url 209"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 208",
      "FeedItemHash": 108,
      "FeedItemID": 19,
      "FeedItemTitle": "title 208",
      "FeedItemUrl": "url 208"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 207",
      "FeedItemHash": 107,
      "FeedItemID": 18,
      "FeedItemTitle": "title 207",
      "FeedItemUrl": "url 207"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 206",
      "FeedItemHash": 106,
      "FeedItemID": 17,
      "FeedItemTitle": "title 206",
      "FeedItemUrl": "url 206"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 205",
      "FeedItemHash": 105,
      "FeedItemID": 16,
      "FeedItemTitle": "title 205",
      "FeedItemUrl": "url 205"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 204",
      "FeedItemHash": 104,
      "FeedItemID": 15,
      "FeedItemTitle": "title 204",
      "FeedItemUrl": "url 204"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 203",
      "FeedItemHash": 103,
      "FeedItemID": 14,
      "FeedItemTitle": "title 203",
      "FeedItemUrl": "url 203"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 202",
      "FeedItemHash": 102,
      "FeedItemID": 13,
      "FeedItemTitle": "title 202",
      "FeedItemUrl": "url 202"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 201",
      "FeedItemHash": 101,
      "FeedItemID": 12,
      "FeedItemTitle": "title 201",
      "FeedItemUrl": "url 201"
    },
    {
      "FeedID": 2,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 200",
      "FeedItemHash": 100,
      "FeedItemID": 11,
      "FeedItemTitle": "title 200",
      "FeedItemUrl": "url 200"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 10",
      "FeedItemHash": 10,
      "FeedItemID": 10,
      "FeedItemTitle": "title 10",
      "FeedItemUrl": "url 10"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 9",
      "FeedItemHash": 9,
      "FeedItemID": 9,
      "FeedItemTitle": "title 9",
      "FeedItemUrl": "url 9"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 8",
      "FeedItemHash": 8,
      "FeedItemID": 8,
      "FeedItemTitle": "title 8",
      "FeedItemUrl": "url 8"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 7",
      "FeedItemHash": 7,
      "FeedItemID": 7,
      "FeedItemTitle": "title 7",
      "FeedItemUrl": "url 7"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 6",
      "FeedItemHash": 6,
      "FeedItemID": 6,
      "FeedItemTitle": "title 6",
      "FeedItemUrl": "url 6"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 5",
      "FeedItemHash": 5,
      "FeedItemID": 5,
      "FeedItemTitle": "title 5",
      "FeedItemUrl": "url 5"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 4",
      "FeedItemHash": 4,
      "FeedItemID": 4,
      "FeedItemTitle": "title 4",
      "FeedItemUrl": "url 4"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 3",
      "FeedItemHash": 3,
      "FeedItemID": 3,
      "FeedItemTitle": "title 3",
      "FeedItemUrl": "url 3"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 2",
      "FeedItemHash": 2,
      "FeedItemID": 2,
      "FeedItemTitle": "title 2",
      "FeedItemUrl": "url 2"
    },
    {
      "FeedID": 1,
      "FeedItemAddedTime": "2000-01-01T00:00:00",
      "FeedItemDescription": "desc 1",
      "FeedItemHash": 1,
      "FeedItemID": 1,
      "FeedItemTitle": "title 1",
      "FeedItemUrl": "url 1"
    }
  ],
  "resultFeeds": [
    {
      "FeedHash": null,
      "FeedID": 10,
      "FeedName": "feed 10",
      "FeedUrl": "url 10"
    },
    {
      "FeedHash": null,
      "FeedID": 9,
      "FeedName": "feed 9",
      "FeedUrl": "url 9"
    },
    {
      "FeedHash": null,
      "FeedID": 8,
      "FeedName": "feed 8",
      "FeedUrl": "url 8"
    },
    {
      "FeedHash": null,
      "FeedID": 7,
      "FeedName": "feed 7",
      "FeedUrl": "url 7"
    },
    {
      "FeedHash": null,
      "FeedID": 6,
      "FeedName": "feed 6",
      "FeedUrl": "url 6"
    },
    {
      "FeedHash": null,
      "FeedID": 5,
      "FeedName": "feed 5",
      "FeedUrl": "url 5"
    },
    {
      "FeedHash": 4,
      "FeedID": 4,
      "FeedName": "feed 4",
      "FeedUrl": "url 4"
    },
    {
      "FeedHash": null,
      "FeedID": 3,
      "FeedName": "feed 3",
      "FeedUrl": "url 3"
    },
    {
      "FeedHash": 2,
      "FeedID": 2,
      "FeedName": "feed 2",
      "FeedUrl": "url 2"
    },
    {
      "FeedHash": 1,
      "FeedID": 1,
      "FeedName": "feed 1",
      "FeedUrl": "url 1"
    }
  ]
}